File: sources/Android/android-communications/library/src/sdk/java/com/polar/sdk/impl/utils/PolarNightlyRechargeUtils.kt
package com.polar.sdk.impl.utils

import com.polar.androidcommunications.api.ble.BleLogger
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.sleep.PolarNightlyRechargeData
import fi.polar.remote.representation.protobuf.NightlyRecovery.PbNightlyRecoveryStatus
import protocol.PftpRequest
import java.time.LocalDate
import java.time.format.DateTimeFormatter

private const val ARABICA_USER_ROOT_FOLDER = "/U/0/"
private const val NIGHTLY_RECOVERY_DIRECTORY = "NR/"
private const val NIGHTLY_RECOVERY_PROTO = "NR.BPB"
private val dateFormatter = DateTimeFormatter.ofPattern("yyyyMMdd")
private const val TAG = "PolarNightlyRechargeUtils"

internal object PolarNightlyRechargeUtils {

    /**
     * Read nightly recharge data for given date range.
     */
    suspend fun readNightlyRechargeData(client: BlePsFtpClient, date: LocalDate): PolarNightlyRechargeData? {
        BleLogger.d(TAG, "readNightlyRechargeData: $date")
        val nightlyRecoveryFilePath = "$ARABICA_USER_ROOT_FOLDER${date.format(dateFormatter)}/$NIGHTLY_RECOVERY_DIRECTORY$NIGHTLY_RECOVERY_PROTO"
        return try {
            val response = client.request(
                PftpRequest.PbPFtpOperation.newBuilder()
                    .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                    .setPath(nightlyRecoveryFilePath)
                    .build()
                    .toByteArray()
            )
            val recoveryStatus = PbNightlyRecoveryStatus.parseFrom(response.toByteArray())
            val recoveryDate = LocalDate.of(
                recoveryStatus.sleepResultDate.year,
                recoveryStatus.sleepResultDate.month,
                recoveryStatus.sleepResultDate.day
            )
            val createdTimestamp = PolarTimeUtils.pbSystemDateTimeToLocalDateTime(recoveryStatus.createdTimestamp)
            val modifiedTimestamp = if (recoveryStatus.hasModifiedTimestamp()) {
                PolarTimeUtils.pbSystemDateTimeToLocalDateTime(recoveryStatus.modifiedTimestamp)
            } else {
                null
            }
            PolarNightlyRechargeData(
                createdTimestamp = createdTimestamp,
                modifiedTimestamp = modifiedTimestamp,
                ansStatus = recoveryStatus.ansStatus,
                recoveryIndicator = recoveryStatus.recoveryIndicator,
                recoveryIndicatorSubLevel = recoveryStatus.recoveryIndicatorSubLevel,
                ansRate = recoveryStatus.ansRate,
                scoreRateObsolete = recoveryStatus.scoreRateOBSOLETE,
                meanNightlyRecoveryRRI = recoveryStatus.meanNightlyRecoveryRRI,
                meanNightlyRecoveryRMSSD = recoveryStatus.meanNightlyRecoveryRMSSD,
                meanNightlyRecoveryRespirationInterval = recoveryStatus.meanNightlyRecoveryRespirationInterval,
                meanBaselineRRI = recoveryStatus.meanBaselineRRI,
                sdBaselineRRI = recoveryStatus.sdBaselineRRI,
                meanBaselineRMSSD = recoveryStatus.meanBaselineRMSSD,
                sdBaselineRMSSD = recoveryStatus.sdBaselineRMSSD,
                meanBaselineRespirationInterval = recoveryStatus.meanBaselineRespirationInterval,
                sdBaselineRespirationInterval = recoveryStatus.sdBaselineRespirationInterval,
                sleepTip = recoveryStatus.sleepTip,
                vitalityTip = recoveryStatus.vitalityTip,
                exerciseTip = recoveryStatus.exerciseTip,
                sleepResultDate = recoveryDate
            )
        } catch (error: Throwable) {
            BleLogger.w(TAG, "Failed to fetch nightly recharge for date: $date, error: $error")
            null
        }
    }
}

File: sources/iOS/ios-communications/Sources/PolarBleSdk/sdk/impl/utils/PolarNightlyRechargeUtils.swift
//  Copyright © 2024 Polar. All rights reserved.

import Foundation

private let ARABICA_USER_ROOT_FOLDER = "/U/0/"
private let NIGHTLY_RECOVERY_DIRECTORY = "NR/"
private let NIGHTLY_RECOVERY_PROTO = "NR.BPB"
private let dateFormat: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyyMMdd"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter
}()
private let TAG = "PolarNightlyRechargeUtils"

internal class PolarNightlyRechargeUtils {
    enum PolarNightlyRechargeError: Error { case missingOrInvalidRecoveryDate }

    static func readNightlyRechargeData(client: BlePsFtpClient, date: Date) async -> PolarNightlyRechargeData? {
        BleLogger.trace(TAG, "readNightlyRechargeData: \(date)")
        let filePath = "\(ARABICA_USER_ROOT_FOLDER)\(dateFormat.string(from: date))/\(NIGHTLY_RECOVERY_DIRECTORY)\(NIGHTLY_RECOVERY_PROTO)"
        let operation = Protocol_PbPFtpOperation.with { $0.command = .get; $0.path = filePath }
        do {
            let response = try await client.request(try operation.serializedBytes())
            let recoveryStatus = try Data_PbNightlyRecoveryStatus(serializedBytes: Data(response))
            guard let recoveryDate = try? PolarTimeUtils.pbDateToDateComponents(pbDate: recoveryStatus.sleepResultDate) else {
                throw PolarNightlyRechargeError.missingOrInvalidRecoveryDate
            }
            let createdTimestamp = try PolarTimeUtils.pbSystemDateTimeToDate(pbSystemDateTime: recoveryStatus.createdTimestamp)
            let modifiedTimestamp = recoveryStatus.hasModifiedTimestamp ? try PolarTimeUtils.pbSystemDateTimeToDate(pbSystemDateTime: recoveryStatus.modifiedTimestamp) : nil
            return PolarNightlyRechargeData(
                createdTimestamp: createdTimestamp, modifiedTimestamp: modifiedTimestamp,
                ansStatus: Float(recoveryStatus.ansStatus), recoveryIndicator: Int(recoveryStatus.recoveryIndicator),
                recoveryIndicatorSubLevel: Int(recoveryStatus.recoveryIndicatorSubLevel), ansRate: Int(recoveryStatus.ansRate),
                scoreRateObsolete: Int(recoveryStatus.scoreRateObsolete), meanNightlyRecoveryRRI: Int(recoveryStatus.meanNightlyRecoveryRri),
                meanNightlyRecoveryRMSSD: Int(recoveryStatus.meanNightlyRecoveryRmssd),
                meanNightlyRecoveryRespirationInterval: Int(recoveryStatus.meanNightlyRecoveryRespirationInterval),
                meanBaselineRRI: Int(recoveryStatus.meanBaselineRri), sdBaselineRRI: Int(recoveryStatus.sdBaselineRri),
                meanBaselineRMSSD: Int(recoveryStatus.meanBaselineRmssd), sdBaselineRMSSD: Int(recoveryStatus.sdBaselineRmssd),
                meanBaselineRespirationInterval: Int(recoveryStatus.meanBaselineRespirationInterval),
                sdBaselineRespirationInterval: Int(recoveryStatus.sdBaselineRespirationInterval),
                sleepTip: recoveryStatus.sleepTip, vitalityTip: recoveryStatus.vitalityTip,
                exerciseTip: recoveryStatus.exerciseTip, sleepResultDate: recoveryDate
            )
        } catch {
            BleLogger.error("readNightlyRechargeData() failed for path: \(filePath), error: \(error)")
            return nil
        }
    }
}
